components/brave_sync/brave_sync_service_impl.cc

```diff
@@ -763,35 +566,145 @@ void BraveSyncServiceImpl::OnResolvedPreferences(const RecordsList& records,
+
+    auto bookmark_record = sync_record->GetBookmark();
+    std::vector<const bookmarks::BookmarkNode*> nodes;
+    bookmark_model_->GetNodesByURL(
+        GURL(bookmark_record.site.location), &nodes);
+
+    std::string object_id;
+    const bookmarks::BookmarkNode* node;
+    for (const auto* i : nodes) {
+      i->GetMetaInfo("object_id", &object_id);
+      if (object_id.empty())
+        continue;
+
+      if (object_id == sync_record->objectId) {
+        node = i;
+        break;
+      }
     }
-    // Abnormal cases
-    if (sync_record->action == jslib::SyncRecord::Action::DELETE && local_id.empty()) {
-      LOG(WARNING) << "received request to delete bookmark which we don't have, ignoring";
-    } else if (sync_record->action == jslib::SyncRecord::Action::CREATE && !local_id.empty()) {
-      LOG(WARNING) << "received request to create bookmark which already exists, ignoring";
-    } else if (sync_record->action == jslib::SyncRecord::Action::UPDATE && local_id.empty()) {
-      LOG(WARNING) << "received request to update bookmark which we don't have, ignoring";
+
+    if (node && sync_record->action == jslib::SyncRecord::Action::UPDATE) {
+      UpdateNode(bookmark_model_, node, sync_record.get());
+      int64_t old_parent_local_id = node->parent()->id();
+      const bookmarks::BookmarkNode* old_parent_node =
+          bookmarks::GetBookmarkNodeByID(bookmark_model_, old_parent_local_id);
+
+      std::string old_parent_object_id;
+      if (old_parent_node) {
+        old_parent_node->GetMetaInfo("object_id", &old_parent_object_id);
+      }
+
+      const bookmarks::BookmarkNode* new_parent_node;
+      if (bookmark_record.parentFolderObjectId != old_parent_object_id) {
+        new_parent_node = FindByObjectId(bookmark_model_,
+                                         bookmark_record.parentFolderObjectId);
+        // TODO - what if new_parent_node doesn't exist yet?
+        DCHECK(new_parent_node);
+      } else {
+        new_parent_node = nullptr;
+      }
+
+      if (new_parent_node) {
+        int64_t index = GetIndex(new_parent_node, bookmark_record);
+        bookmark_model_->Move(node, new_parent_node, index);
+      }
+    } else if (node && sync_record->action == jslib::SyncRecord::Action::DELETE) {
+      bookmark_model_->Remove(node);
+    } else if (!node) {
+      // TODO (make sure there isn't an existing record with the objectId)
+
+      const bookmarks::BookmarkNode* parent_node =
+          FindByObjectId(bookmark_model_, bookmark_record.parentFolderObjectId);
+
+      if (!parent_node) {
+        if (!bookmark_record.order.empty() &&
+            bookmark_record.order.at(0) == '2') {
+          // mobile generated bookmarks go in the mobile folder so they don't get
+          // so we don't get m.xxx.xxx domains in the normal bookmarks
+          parent_node = bookmark_model_->mobile_node();
+        } else {
+          parent_node = bookmark_model_->other_node();
+        }
```

Comment:
will have to deal with Bookmarks Bar

Reply:
yes, but it's unclear how we should deal with that to make it consistent. We can't use the object_id because it will be different on every device. Maybe a variation on the existing flag to indicate that is is in the bookmarks bar vs isn't?

Reply:
we have `Bookmark::hideInToolbar` field, which should be set to false if bookmark appears on toolbar. so if bookmark sync record has hideInToolbar==false and parentObjectId empty it should be child of a toolbar node. Nested children will  have parentObjectId non-empty.

Reply:
the flag just seems backwards to me. Shouldn't it be showInToolbar instead of hideInToolbar?

Reply:
for me it was also hard to understand meaning of `hideInToolbar` . I found the explanation maybe oce all bookmarks were in toolbar and then some could be placed not in toolbar, so the flag had such name.
It goes from  https://github.com/brave/sync/blob/staging/lib/api.proto#L50